sql/src/main/java/io/crate/operation/projectors/BatchIteratorBackpressureExecutor.java

```diff
@@ -55,145 +54,120 @@
+        int inFlight = inFlightExecutions.decrementAndGet();
+        assert inFlight >= 0 : "Number of in-flight executions must not be negative";
 
-    private BiConsumer<R, Throwable> createBulkExecutionCompleteListener(BatchIterator batchIterator) {
-        return (r, t) -> {
-            int inFlight = inFlightExecutions.decrementAndGet();
-            if (inFlight == 0 && lastBulkScheduledToExecute && computeFinalResult.compareAndSet(false, true)) {
-                // all bulks are complete, close iterator and complete executionFuture
-                batchIterator.close();
-                if (t == null) {
-                    resultFuture.complete(null);
-                } else {
-                    resultFuture.completeExceptionally(t);
-                }
-            } else {
-                safeConsumeIterator(batchIterator);
+        if (consumptionFinished) {
+            if (inFlight == 0) {
+                setResult(failure);
             }
-        };
+            // else: waiting for other async-operations to finish
+        } else {
+            consumeIterator();
+        }
     }
 
-    /**
-     * Consumes the iterator only if there isn't another thread already consuming it.
-     */
-    private void safeConsumeIterator(BatchIterator batchIterator) {
-        if (collectingEnabled.compareAndSet(false, true)) {
-            unsafeConsumeIterator(batchIterator);
+    private void setResult(Throwable failure) {
+        batchIterator.close();
+        if (failure == null) {
+            resultFuture.complete(null);
+        } else {
+            resultFuture.completeExceptionally(failure);
         }
     }
 
     /**
-     * Consumes the iterator without guarding against concurrent access to the iterator.
+     * Consumes the rows from the BatchIterator and invokes {@link #executeFunction} every {@link #bulkSize} rows.
+     * This loop continues until either:
      *
-     * !! THIS SHOULD ONLY BE CALLED BY THE loadNextBatch COMPLETE LISTENER !!
+     *  - The BatchIterator has been fully consumed.
+     *  - {@link #pauseConsumption} returns true; In this case a scheduler is used to re-schedule the consumption
+     *    after a throttle-delay
+     *
+     * Each (async) operation which is executed *could* either set the result on {@link #resultFuture} or re-resume
+     * consumption.
+     * To make sure only a single consumer thread is active at a time a {@link #semaphore} is used.
      */
-    private void unsafeConsumeIterator(BatchIterator batchIterator) {
-        if(resultFuture.isDone()) {
-            // if the execution future was completed (most likely failed or killed) we stop consuming
+    private void consumeIterator() {
+        if (semaphore.tryAcquire() == false) {
             return;
         }
         Row row = RowBridging.toRow(batchIterator.rowData());
         try {
-            while (true) {
+            while (batchIterator.moveNext()) {
```

Comment:
This breaks the batches completely. 
The moveNext call was in a `while(true)` loop because when coming back from a scheduled bulk you want to execute it (ie. go through `tryExecuteBulk` again) and THEN resume consumption (ie `moveNext` and  build the next bulk)
With current code after the first bulk is rescheduled and comes back into `consumeIterator` it will move the iterator to the next value and it will increment the `indexInBulk` to `bulkSize`+1 (because a bulk was rescheduled when `indexInBulk` == `bulkSize`). This means ` if (indexInBulk == bulkSize) ` will never be true and we'll end up with a massive bulk which will get executed when the iterator has no more items